Do not count hidden categories in tracking budgets + cleanup table math

lelemm opened the pull request:
This changes the budget table calculations to exclude hidden categories from totals.  To test, hide and unhide categories and groups and compare the group and budget totals.

This also changes the summation for balance and spending to include all amounts that month regardless of a category balance being carried forward.  The balance will still carry to the next month for now.

fixes #4116
resolves #2400

youngcw commented:
Working for everything except income categories.  Hiding/unhiding those is broken

youngcw commented on packages/loot-core/src/server/budget/base.ts:
```diff
@@ -227,6 +251,28 @@ function handleCategoryChange(months, oldValue, newValue) {
+    const id = newValue.id;
+    const groupId = newValue.cat_group;
+    //TODO: Make this work
+    const isIncome = db.first(
```
Add a functional db call here (or some other way to get the correct is_income value), and it should work.

youngcw commented:
@lelemm @psybers Want to test this out?  I think its all working as expected now

youngcw commented:
/update-vrt

youngcw commented:
Looks like I broke the envelope budget math

lelemm commented:
@youngcw I moved the new logic to `./report.ts` and moved the old logic to `./envelope.ts` this way we can avoid changing envelop logic.
I did some testing with report. Looks like working. Still need to test envelope

lelemm commented:
/update-vrt

youngcw commented on packages/loot-core/src/server/budget/envelope.ts:
```diff
@@ -178,3 +183,186 @@ export function createBudget(meta, categories, months) {
+  }
+}
+
+export function createCategoryGroup(group, sheetName) {
```
This should probably be at the top like the other cell creation functions.

lelemm replied:
done

lelemm replied:
done

youngcw commented on packages/loot-core/src/server/budget/envelope.ts:
```diff
@@ -178,3 +183,186 @@ export function createBudget(meta, categories, months) {
+    run: sumAmounts,
+  });
+
+  if (!group.is_income || getBudgetType() !== 'rollover') {
```
we shouldn't need to care about budget type in this file

youngcw commented on packages/loot-core/src/server/budget/report.ts:
```diff
@@ -115,3 +117,227 @@ export function createSummary(groups, categories, sheetName) {
+  const budgetType = getBudgetType();
+
+  function addDeps(sheetName, groupId, catId, isIncome = null) {
+    if (getBudgetType() === 'rollover' || isIncome) {
```
again, shouldn't need to check which budget type

youngcw commented on packages/loot-core/src/server/budget/report.ts:
```diff
@@ -115,3 +117,227 @@ export function createSummary(groups, categories, sheetName) {
+  }
+}
+
+export function createCategoryGroup(group, sheetName) {
```
should be near top

lelemm replied:
done

youngcw commented on packages/loot-core/src/server/budget/report.ts:
```diff
@@ -115,3 +117,227 @@ export function createSummary(groups, categories, sheetName) {
+  }
+
+  function removeDeps(sheetName, groupId, catId, isIncome = null) {
+    if (getBudgetType() === 'rollover' || isIncome) {
```
extra check again

lelemm replied:
done

lelemm commented:
I didn't clean the code logic yet.

youngcw commented:
/update-vrt

youngcw commented:
/update-vrt

youngcw commented:
Found a bug.  If a category or group is hidden before changing to tracking budget, it isn't excluded from the totals.  If clears itself up after showing/hiding.  We probably need to initially exclude hidden cells in the way that the original PR was doing.

lelemm commented:
lol. we both fixed it @youngcw

lelemm commented:
I merged your changes over mine

youngcw commented:
We need to do something about group balances.  Currently they still consider the carried over balances, but the budget balance does not since it pulls from the spending field.

youngcw commented:
I made all the totals ignore if something is carried forward or not.  I think that makes the most sense.  The balance for each category will still work with carrying forward amounts, but all the totals sum up the table.

kevingatera commented:
@lelemm 
If I understand correctly, the goal of the PR would be to, then I will bring some tests to:
- Exclude hidden categories from group totals in Report Budget
- Exclude hidden category groups from budget totals in Report Budget
- Include hidden categories in group totals for Rollover Budget
- Include hidden category groups in budget totals for Rollover Budget

If so, that's what my contribution is going to be, let me know of how you'd like to get it, as a PR against your branch or as a commit to this one

lelemm commented:
> * Include hidden categories in group totals for Rollover Budget
> * Include hidden category groups in budget totals for Rollover Budget

isn't this already happening?

kevingatera commented:
Yes those last 2 are already happening but untested afaik. I will open a PR against your branch @lelemm

youngcw commented:
@lelemm if you think this is ready then we can call it good and merge it
